File: Offline/OfflinePlayer/objc/OfflinePlayer/Table View Cells/VideoTableViewCell.m
//
//  VideoTableViewCell.m
//  OfflinePlayer
//
//  Created by Jeremy Blaker on 10/28/19.
//  Copyright © 2019 Brightcove. All rights reserved.
//

#import "VideoTableViewCell.h"

@import BrightcovePlayerSDK;

@interface VideoTableViewCell ()

@property (nonatomic, weak) BCOVVideo *video;

@property (nonatomic, weak) IBOutlet UILabel *titleLabel;
@property (nonatomic, weak) IBOutlet UILabel *detailLabel;
@property (nonatomic, weak) IBOutlet UIProgressView *progressView;
@property (nonatomic, weak) IBOutlet UIImageView *statusButtonImageView;
@property (nonatomic, weak) IBOutlet UIImageView *thumbnailImageView;

- (IBAction)downloadButtonWasPressed:(id)sender;

@end

// Custom cell implementation to arrange
// text and images more carefully.
// Also adds a download status image.
@implementation VideoTableViewCell : UITableViewCell

- (void)awakeFromNib
{
    [super awakeFromNib];
    [self cleanup];
}

- (void)prepareForReuse
{
    [super prepareForReuse];
    [self cleanup];
}

- (void)cleanup
{
    self.titleLabel.text = nil;
    self.detailLabel.text = nil;
    self.thumbnailImageView.image = nil;
    self.statusButtonImageView.image = nil;
}

- (IBAction)downloadButtonWasPressed:(id)sender
{
    if ([self.delegate respondsToSelector:@selector(downloadButtonTappedForVideo:)])
    {
        [self.delegate downloadButtonTappedForVideo:self.video];
    }
}

- (void)setupWithStreamingVideo:(BCOVVideo *)video estimatedDownloadSize:(double)downloadSize thumbnailImage:(UIImage *)thumbnailImage videoState:(VideoState)videoState
{
    self.video = video;
    
    self.progressView.hidden = YES;
    
    [self setupTitleLabel];
    
    NSString *detailString = self.video.properties[kBCOVVideoPropertyKeyDescription];
    if ((detailString == nil) || (detailString.length == 0))
    {
        detailString = self.video.properties[kBCOVVideoPropertyKeyReferenceId] ?: @"nil";
    }

    NSNumber *durationNumber = self.video.properties[kBCOVVideoPropertyKeyDuration];
    // raw duration is in milliseconds
    int duration = durationNumber.intValue / 1000;

    self.detailLabel.text = [NSString stringWithFormat:@"%d sec / %0.2f MB\n%@", duration, downloadSize, detailString];

    self.thumbnailImageView.image = thumbnailImage = thumbnailImage ?: [UIImage imageNamed:@"bcov"];
    
    [self setStateImage:videoState];
}

- (void)setupWithOfflineVideo:(BCOVVideo *)video offlineStatus:(BCOVOfflineVideoStatus *)offlineStatus downloadSize:(double)downloadSize
{
    self.video = video;
    
    self.progressView.hidden = NO;
    
    [self setupTitleLabel];
    
    NSString *detailString = video.properties[kBCOVVideoPropertyKeyDescription];
    if ((detailString == nil) || (detailString.length == 0))
    {
        detailString = video.properties[kBCOVVideoPropertyKeyReferenceId] ?: @"";
    }

    // Detail text is two lines consisting of:
    // "duration in seconds / actual download size)"
    // "reference_id"
    NSNumber *durationNumber = video.properties[kBCOVVideoPropertyKeyDuration];
    // raw duration is in milliseconds
    int duration = durationNumber.intValue / 1000;
    NSString *twoLineDetailString;
    
    if (offlineStatus.downloadState == BCOVOfflineVideoDownloadStateCompleted)
    {
        // Use Kilobytes if the measurement is too small
        if (downloadSize < 0.5)
        {
            double kilobytes = downloadSize * 1000.0;
            twoLineDetailString = [NSString stringWithFormat:@"%d sec / %0.2f KB\n%@",
                                   duration, kilobytes,
                                   detailString];
        }
        else
        {
            twoLineDetailString = [NSString stringWithFormat:@"%d sec / %0.2f MB\n%@",
                                   duration, downloadSize,
                                   detailString];
        }
    }
    else
    {
        // download not complete: skip the download size
        twoLineDetailString = [NSString stringWithFormat:@"%d sec / %@ MB\n%@",
                               duration, @"--",
                               detailString];
    }
    
    self.detailLabel.text = twoLineDetailString;
    
    NSString *thumbnailPathString = video.properties[kBCOVOfflineVideoThumbnailFilePathPropertyKey];
    UIImage *thumbnailImage = [UIImage imageWithContentsOfFile:thumbnailPathString];

    // Set up the image view
    // Use a default image if the cached image is not available
    self.thumbnailImageView.image = thumbnailImage ?: [UIImage imageNamed:@"bcov"];
    
    if (offlineStatus == nil)
    {
        [self setStateImage:VideoStateOnlineOnly];
    }
    else
    {
        switch (offlineStatus.downloadState)
        {
            case BCOVOfflineVideoDownloadLicensePreloaded:
            case BCOVOfflineVideoDownloadStateRequested:
            case BCOVOfflineVideoDownloadStateTracksRequested:
            case BCOVOfflineVideoDownloadStateDownloading:
            case BCOVOfflineVideoDownloadStateTracksDownloading:
                [self setStateImage:VideoStateDownloading];
                break;
            case BCOVOfflineVideoDownloadStateSuspended:
            case BCOVOfflineVideoDownloadStateTracksSuspended:
                [self setStateImage:VideoStatePaused];
                break;
            case BCOVOfflineVideoDownloadStateCancelled:
            case BCOVOfflineVideoDownloadStateTracksCancelled:
                [self setStateImage:VideoStateCancelled];
                break;
            case BCOVOfflineVideoDownloadStateCompleted:
            case BCOVOfflineVideoDownloadStateTracksCompleted:
                [self setStateImage:VideoStateDownloaded];
                break;
            case BCOVOfflineVideoDownloadStateError:
            case BCOVOfflineVideoDownloadStateTracksError:
                [self setStateImage:VideoStateError];
                break;
        }
    }
    
    self.progressView.progress = offlineStatus.downloadPercent / 100;
}

- (void)setupTitleLabel
{
    self.titleLabel.text = self.video.properties[kBCOVVideoPropertyKeyName];
    self.titleLabel.textColor = (self.video.usesFairPlay ? [UIColor colorWithRed:0.75 green:0.0 blue:0.0 alpha:1.0] : UIColor.blackColor);
}

- (void)setStateImage:(VideoState)state
{
    UIImage *newImage = nil;
    switch (state)
    {
        case VideoStateOnlineOnly: // nothing
        {
            break;
        }
        case VideoStateDownloadable:
        {
            newImage = [UIImage imageNamed:@"download"];
            break;
        }
        case VideoStateDownloading:
        {
            newImage = [UIImage imageNamed:@"inprogress"];
            break;
        }
        case VideoStatePaused:
        {
            newImage = [UIImage imageNamed:@"paused"];
            break;
        }
        case VideoStateDownloaded:
        {
            newImage = [UIImage imageNamed:@"downloaded"];
            break;
        }
        case VideoStateCancelled:
        {
            newImage = [UIImage imageNamed:@"cancelled"];
            break;
        }
        case VideoStateError:
        {
            newImage = [UIImage imageNamed:@"error"];
            break;
        }
    }

    self.statusButtonImageView.image = newImage;
}

@end

File: Offline/OfflinePlayer/swift/OfflinePlayer/TableCells/VideoTableViewCell.swift
//
//  VideoTableViewCell.swift
//  OfflinePlayer
//
//  Copyright © 2019 Brightcove, Inc. All rights reserved.
//

import UIKit
import BrightcovePlayerSDK

class VideoTableViewCell: UITableViewCell {

    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var detailsLabel: UILabel!
    @IBOutlet weak var statusButton: UIButton!
    @IBOutlet weak var thumbnailImageView: UIImageView!
    @IBOutlet weak var progressView: UIProgressView! {
        didSet {
            progressView.isHidden = true
        }
    }
    
    weak var video: BCOVVideo?
    weak var delegate: VideoTableViewCellDelegate?
    
    override func awakeFromNib() {
        super.awakeFromNib()
        cleanup()
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        cleanup()
    }
    
    private func cleanup() {
        titleLabel.text = nil
        detailsLabel.text = nil
        thumbnailImageView.image = nil
        statusButton.setImage(nil, for: .normal)
    }
    
    private func updateStatusButtonImage(_ state: VideoState) {
        
        var stateImage: UIImage?
        
        switch state {
        case .OnlineOnly:
                break;
        case .Downloadable:
            stateImage = UIImage(named: "download")
        case .Downloading:
            stateImage = UIImage(named: "inprogress")
        case .Paused:
            stateImage = UIImage(named: "paused")
        case .Downloaded:
            stateImage = UIImage(named: "downloaded")
        case .Cancelled:
            stateImage = UIImage(named: "cancelled")
        case .Error:
            stateImage = UIImage(named: "error")
        }
        
        DispatchQueue.main.async {
            self.statusButton.setImage(stateImage, for: .normal)
        }
        
    }
    
    private func handleDownloadState(offlineStatus: BCOVOfflineVideoStatus) {
        
        switch offlineStatus.downloadState {
            case .licensePreloaded,
                 .stateRequested,
                 .stateTracksRequested,
                 .stateDownloading,
                 .stateTracksDownloading:
            updateStatusButtonImage(.Downloading)
            case .stateSuspended,
                 .stateTracksSuspended:
            updateStatusButtonImage(.Paused)
            case .stateCancelled,
                 .stateTracksCancelled:
            updateStatusButtonImage(.Cancelled)
            case .stateCompleted,
                 .stateTracksCompleted:
            updateStatusButtonImage(.Downloaded)
            case .stateError,
                 .stateTracksError:
            updateStatusButtonImage(.Error)
        }
        
    }
    
    func setup(withStreamingVideo video: BCOVVideo, estimatedDownloadSize: Double, thumbnailImage: UIImage?, videoState: VideoState) {
        
        self.video = video
        
        setupTitleLabel(withVideo: video)
        
        let detailString = getDetailString(forVideo: video)
        
        // Detail text is two lines consisting of:
        // "duration in seconds / estimated download size)"
        // "reference_id"
        if let durationNumber = video.properties["duration"] as? NSNumber {
            // raw duration is in milliseconds
            let duration = durationNumber.intValue / 1000
            let twoLineDetailString = "\(duration) sec / \(formattedSizeString(filesize: estimatedDownloadSize)) MB \n\(detailString)"
            
            detailsLabel.text = twoLineDetailString
        }
        
        // Use cached thumbnail image for dipslay
        thumbnailImageView.image = thumbnailImage ?? UIImage(named: "bcov")
        
        updateStatusButtonImage(videoState)
        
        statusButton.isUserInteractionEnabled = true
    }
    
    func setup(withOfflineVideo video: BCOVVideo, offlineStatus: BCOVOfflineVideoStatus, downloadSize: Double) {
        
        setupTitleLabel(withVideo: video)
        
        let detailString = getDetailString(forVideo: video)
        var twoLineDetailString: String?
        
        // Detail text is two lines consisting of:
        // "duration in seconds / actual download size)"
        // "reference_id"
        if let durationNumber = video.properties["duration"] as? NSNumber {
        
            let duration = durationNumber.intValue / 1000
            
            if offlineStatus.downloadState == .stateCompleted {
                
                // download complete: show the downloaded video size
                
                // Use Kilobytes if the measurement is too small
                if downloadSize < 0.5 {
                    let kilobytes = downloadSize * 1000
                    twoLineDetailString = "\(duration) sec / \(formattedSizeString(filesize: kilobytes)) KB\n\(detailString)"
                } else {
                    twoLineDetailString = "\(duration) sec / \(formattedSizeString(filesize: downloadSize)) MB\n\(detailString)"
                }
                
                
            } else {
                // download not complete: skip the download size
                twoLineDetailString = "\(duration) sec / -- MB\n\(detailString)"
            }
            
            detailsLabel.text = twoLineDetailString
            
        }
        
        // Set the thumbnail image
        if let thumbnailPathString = video.properties[kBCOVOfflineVideoThumbnailFilePathPropertyKey] as? String {
            let thumbnailImage = UIImage(contentsOfFile: thumbnailPathString)
            thumbnailImageView?.image = thumbnailImage ?? UIImage(named: "bcov")
        }
        
        handleDownloadState(offlineStatus: offlineStatus)
        
        progressView.isHidden = false
        updateProgressView(withPercentage: offlineStatus.downloadPercent)
        
        statusButton.setImage(UIImage(named: "downloaded"), for: .normal)
        statusButton.isUserInteractionEnabled = false
    }
    
    private func updateProgressView(withPercentage percentage: CGFloat) {
        self.progressView.setProgress(Float(percentage / 100.0), animated: true)
    }
    
    private func formattedSizeString(filesize: Double) -> String {
        return String(format: "%0.2f", filesize)
    }
    
    private func getDetailString(forVideo video: BCOVVideo) -> String {
        guard let detailString = video.properties["description"] as? String else {
            return video.properties["reference_id"] as? String ?? ""
        }
        
        return detailString
    }
    
    private func setupTitleLabel(withVideo video: BCOVVideo) {
        titleLabel.text = video.properties["name"] as? String
        
        // Use red label to indicate that the video is protected with FairPlay
        titleLabel.textColor = video.usesFairPlay ? UIColor(red: 0.75, green: 0.0, blue: 0.0, alpha: 1.0) : UIColor.black
    }
    
    @IBAction private func downloadButtonPressed(_ button: UIButton) {
        if let video = video {
            delegate?.performDownload(forVideo: video)
        }
    }

}

protocol VideoTableViewCellDelegate: class {
    func performDownload(forVideo: BCOVVideo)
}
